### PR TITLE
Remove delete mapping API

### LINKED ISSUE #8877: Remove the ability to delete mappings
Currently, a user can delete a `type` entirely, which deletes the type, all documents of that type, and removes the associated fields.  Unfortunately, if any documents of that type have already been indexed, then the fields are part of the index.  Adding fields with the same name but a different data type will cause conflicts later on. (See https://github.com/elasticsearch/elasticsearch/issues/8530#issuecomment-64393335)

We should remove the ability to delete types/mappings.

Relates to #8870

### PR DESCRIPTION
**Pull request by rjernst**
Deleting a type from an index is inherently dangerous because
the type can be recreated with new mappings which may conflict
with existing segments still using the old mappings. This
removes the ability to delete a type (similar to how deleting
fields within a type is not allowed, for the same reason).

closes #8877

### CODE REVIEW

**Comment by martijnvg** [comment]
Can you add an entry to the migrate_2_0.asciidoc about the fact that this api has been deleted? Other than that LGTM.

**Comment by rjernst** [comment]
@martijnvg Sure, added a short message to the docs.

**Comment by martijnvg** [comment]
LGTM

**Comment by jpountz** [comment]
LGTM

File: src/main/java/org/elasticsearch/river/RiversService.java [hunk 1]
```diff
@@ -214,37 +211,6 @@ public void riverClusterChanged(RiverClusterChangedEvent event) {
...
-                            @Override
-                            public void onFailure(Throwable e) {
-                                logger.debug("failed to (double) delete river [{}] content", e, riverName.name());
-                            }
-                        });
-                    } catch (IndexMissingException e) {
-                        // all is well, the _river index was deleted
-                    } catch (Exception e) {
-                        logger.warn("unexpected failure when trying to verify river [{}] deleted", e, riverName.name());
-                    }
```
Comment (javanna) [inline src/main/java/org/elasticsearch/river/RiversService.java]: I wonder what effect this change has on rivers... not sure...maybe next step would be to remove rivers code from master too? :)

**Comment by javanna** [comment]
ping @spalger when this gets pushed you'll have to regenerate the api for the js client otherwise the references to the non existing docs page might break docs publish process. Thanks!

File: src/main/java/org/elasticsearch/river/RiversService.java (hunk 1 above)
Comment (javanna) [inline src/main/java/org/elasticsearch/river/RiversService.java]: @kimchy WDYT?

**Comment by s1monw** [comment]
LGTM too

File: src/main/java/org/elasticsearch/river/RiversService.java (hunk 1 above)
Comment (dadoonet) [inline src/main/java/org/elasticsearch/river/RiversService.java]: I agree with @javanna here. River plugins can't be use anymore as you can not remove a `_river` instance.

To remove a `_river`, we were doing:

```
DELETE _river/rivername
```

What can we do now?

FYI: this PR is the reason why we have all river plugins failing in Jenkins now.

Comment (rjernst) [inline src/main/java/org/elasticsearch/river/RiversService.java]: @dadoonet I have a PR up adding back deletion, but without deleting the type.  See #10251

Comment (dadoonet) [inline src/main/java/org/elasticsearch/river/RiversService.java]: I was wrong. To remove a _river we should run:

       DELETE _river/rivername/_meta

**Comment by aparo** [comment]
How to manage if an existing mapping is wrong?
Usually people deletes it and recreate it.
He can force deleting the documents with a scan first. Optimize the index removing deleted docs and then remove the mapping.
Otherwise people must recreate/reindex indices every time they are a broken mapping!
How we manage generic optimization of deleting empty mappings in indices?

**Comment by clintongormley** [comment]
@aparo There are a number of reasons why reindexing is a better approach:
* Even after an optimize, old fields can still exist in the fields info which can lead to corrupt data.
* Fields are going to be stored at index level (with some per-type settings, eg `copy_to`), so you'd have to be sure that fields aren't being used across multiple types.
* Doing an optimize could involve rewriting all of your segments anyway, and stopping indexing while you do it.
* We plan on supporting reindexing directly in Elasticsearch, which will make it easier to just rewrite the index then switch aliases when it is ready.

**Pull request merged**